### PR TITLE
feat(session): add the trajectory-tree data model

### PR DESCRIPTION
**Pull request by guapisolo**
> **Depends on:** #2123
> Stacked on the parent PR's branch. Review / merge the parent first.

## Summary

Add an append-only trajectory forest with deterministic attach-point search.

## Motivation

Session v2 must retain multiple complete trajectories instead of forcing retries and alternative continuations into one linear history.

## Usage

Use `SessionTree.create_node(parent, ...)` to append a root or child generation, then call `find_attach_point(request_messages)` to select the deepest matching path.

## Design Notes

- **Key choices:** Every node stores a full root-to-node token snapshot.
- Commit sequence `seq` is the deterministic ordering key.
- `MAX_NODES=1024` bounds one session forest.

## Verification

- **Tests added:** `test_tree_trajectory.py` covers the trajectory-tree data-model contract.

## Review Focus

- Scrutinize `SessionTree.find_attach_point` for deepest-prefix and latest-sequence tie-breaking.
- Scrutinize `TrajectoryNode.completion_span` as the loss-mask boundary carried into later layers.

### CODE REVIEW

File: miles/rollout/session/v2/tree_trajectory.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            ):
+                i += 1
+            best_overlap = max(best_overlap, offset + i)
+            if i < len(delta):
+                return  # partial delta: this node (and its subtree) is not a candidate
+            matched = offset + len(delta)
+            if matched > best_matched or (matched == best_matched and best is not None and node.seq > best.seq):
+                best, best_matched = node, matched
+            for child in node.children:
+                visit(child, matched)
```
Comment (Shi-Dong) [inline miles/rollout/session/v2/tree_trajectory.py]: Won't this result in a recursion-limit exceeded error if there are too many nodes?

File: miles/rollout/session/v2/tree_trajectory.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    ) -> TrajectoryNode:
+        if len(self.nodes) >= MAX_NODES:
+            raise ValueError(
+                f"node cap reached ({MAX_NODES}): the session cannot branch or extend "
+                f"further — this almost always means the harness is not replaying "
+                f"history verbatim"
+            )
+        node = TrajectoryNode(
+            delta_messages=list(delta_messages),
+            token_ids=token_ids,
```
Comment (Shi-Dong) [inline miles/rollout/session/v2/tree_trajectory.py]: Why not `list(token_ids)`?

File: miles/rollout/session/v2/tree_trajectory.py (hunk 1 above)
Comment (guapisolo) [reply miles/rollout/session/v2/tree_trajectory.py]: Fixed in `ca5ab7e`. I replaced recursive attach-point traversal with an explicit stack, so a valid `MAX_NODES=1024` chain no longer depends on Python's recursion limit. I also added a regression test that lowers the recursion limit below the legal tree depth.

File: miles/rollout/session/v2/tree_trajectory.py (hunk 2 above)
Comment (guapisolo) [reply miles/rollout/session/v2/tree_trajectory.py]: Fixed in `ca5ab7e`. `create_node` now stores `list(token_ids)`, so later mutations to the caller's list cannot change the node's root-to-node token snapshot. Since token IDs are integers, a shallow list copy is sufficient. I also added a regression test for this ownership boundary.

**Review by Shi-Dong — APPROVED** [review]
LGTM!
